packages/sdk/src/libs/clients/lsp/client.ts

```diff
@@ -0,0 +1,103 @@
+import { LongShortPair__factory, LongShortPair } from "@uma/core/contract-types/ethers";
+import type { SignerOrProvider, GetEventType } from "../..";
+import { Event } from "ethers";
+import { Balances } from "../../utils";
+
+export type Instance = LongShortPair;
+
+export type TokensCreated = GetEventType<Instance, "TokensCreated">;
+export type TokensRedeemed = GetEventType<Instance, "TokensRedeemed">;
+export type ContractExpired = GetEventType<Instance, "ContractExpired">;
+export type PositionSettled = GetEventType<Instance, "PositionSettled">;
+
+export function connect(address: string, provider: SignerOrProvider): Instance {
+  return LongShortPair__factory.connect(address, provider);
+}
+
+export interface EventState {
+  sponsors?: string[];
+  longs?: Balances;
+  shorts?: Balances;
+  collateral?: Balances;
+  expired?: boolean;
+  expiredBy?: string;
+}
```

Comment:
in my head, if theres a single deviation from this (ie a property that has to be optional because it cant be defaulted) then that is the same as all parameters being optional, since now the user will always need the code anyways to deal with a special case. im not confident we can always set defaults, so in that case allow any property to be optional.  perhaps im wrong in my extremism, but i find special cases to be really distasteful.  I'm open to reconsidering this though

Reply:
i was just exploring a bit if setting defaults makes sense. This is worth considering, if we by default set expired to false, this is misleading to anyone who is querying events after an event which would have set this to true. the expired event is set to false, but otherwise it would be undefined or would not exist if there was no default. IMO it not existing makes more sense, because you know that you cant know the expired state of the contract based on the events you have.  If it defaults to false, you may inadvertently save the expired state as false, even though you previously had saved it as true.

Reply:
I kinda get what you're saying, but I don't really see type guarantees as all-or-nothing. I think you should guarantee as much as is possible via the type system so that you can reduce the verbosity of runtime checks. Runtime checks add branches and failure points which makes code more difficult to reason about and more verbose. If you pass this type to more than one place that needs to read the value, you'll need to check against undefined _at every point of use_.

For most of the more complex types, I think the undefined case is adding more degrees of freedom than you need. For example, an optional array and an empty array are saying the same thing: nothing has been added to the array yet. Having both states adds complexity where it's not needed. Similar for a balances mapping object. An empty set of balances is the same thing as undefined: nothing as been added yet. It seems just simpler for the type system and for the downstream user of these objects that the number of distinct cases in the code match the number of states that the object can be in.

As for the specific case of the expired boolean, if it's never set to false explicitly, then its two values are undefined and true, right? I assume this because there is no event that tells it to be false, there is only an event to tells it to be true. So by definition, your best resolution is undefined or true. Is my reasoning off on this more subtle point?

We don't need to necessarily change anything here, but I think defaulting to making everything optional if _anything_ is optional removes a lot of power from the type enforcement system. You get a lot less "for free".

Thoughts? We can also continue this discussion outside of this PR, because I don't think this should block anything!

Reply:
yea lets continue this discussion. i can add a refactor task for this, and would like to apply it to all clients if we decide to do this